packages/outstatic/src/components/MetadataBuilder/index.tsx

```diff
@@ -0,0 +1,230 @@
+
+  const { repoOwner, repoSlug, repoBranch, contentPath, monorepoPath } =
+    useContext(OutstaticContext)
+
+  const rootPath = [monorepoPath, contentPath].filter(Boolean).join('/')
+
+  const { data } = useGetFileInformationQuery({
+    variables: {
+      owner: repoOwner,
+      name: repoSlug,
+      expression: `${repoBranch}:${rootPath}`
+    },
+    skip: !rebuild
+  })
+
+  const files = useMemo(() => {
+    // strip object freeze to work with mutable data
+    // https://github.com/apollographql/apollo-client/issues/5987#issuecomment-590938556
+    const o = JSON.parse(
+      JSON.stringify(data?.repository?.object ?? {})
+    ) as DeepNonNullable<GetFileInformationQuery>['repository']['object']
+
+    const output: FileData[] = []
+    const queue = 'entries' in o ? o.entries ?? [] : []
+
+    while (queue.length > 0) {
+      const next = queue.pop()
+      if (next?.object?.__typename === 'Tree') {
+        // subdir - add entries to queue
+        queue.push(...(next.object.entries ?? []))
+      } else if (
+        next?.object?.__typename === 'Blob' &&
+        isIndexable(next.path)
+      ) {
+        // file - add to output
+        output.push({
+          path: next.path,
+          oid: `${next.object.oid}`,
+          commit: hashFromUrl(`${next.object.commitUrl}`)
+        })
+      }
+    }
+    // clone out of Apollo's forced read-only
+    return output
+  }, [data])
+
+  // using useEffect ensures we run a single processing loop
+  useEffect(() => {
+    const takeAndProcess = async (o: FileData) => {
+      const res = await client.query<DocumentQuery, DocumentQueryVariables>({
+        query: DocumentDocument,
+        variables: {
+          owner: repoOwner,
+          name: repoSlug,
+          filePath: `${repoBranch}:${o.path}`
+        }
+      })
+
+      if (res.data.repository?.object?.__typename === 'Blob') {
+        const m = matter(res.data.repository.object.text ?? '')
+        const state = MurmurHash3(res.data.repository.object.text ?? '')
+        const fmd: Partial<OutstaticSchema> = {
+          ...m.data,
+          slug:
+            m.data.slug ?? o.path.replace(/^.+\/(.+)\.(md|mdoc|mdx)?/, '$1'),
+          __outstatic: {
+            commit: o.commit,
+            hash: `${state.result()}`,
+            path: o.path
+          }
+        }
+        return fmd
+      }
+
+      return undefined
+    }
+
+    const fn = async () => {
+      setTotal(Math.max(files.length, 1))
+
+      const chunkSize = 5 // TODO move to constants
+      const queue = chunk(files, chunkSize)
+      const docs: Record<string, unknown>[] = []
+      const pendingOid = fetchOid()
+
+      // process in chunks
+      while (queue.length > 0) {
+        const next = queue.pop()
+        if (!next) continue
+        const all = Promise.allSettled(
+          next.map(async (fd) => {
+            const meta = await takeAndProcess(fd)
+            docs.push({
+              ...meta,
+              category: fd.path
```

Comment:
> Any reason why we are going with category here instead of collection?

Nope. Just bad mental model for what to call the folders on my part. I'll be updating to `collection`.

> I've tested the GraphQL query to see the response for the path. I wonder if we should remove the rootPath from the filepath github gives us when saving the metadata.

I really want to. Unfortunately, we don't expose any of those `process.env` values to the client. We build the metadata DB client side so that it's easier to do concurrency & provide progress. We could update `OST_MONOREPO_PATH` and `OST_CONTENT_PATH` to `NEXT_PUBLIC_OST_MONOREPO_PATH` and `NEXT_PUBLIC_OST_CONTENT_PATH` but that seemed beyond the scope of the current WIP. If you think this could be a good direction, I'll add this to my TODO list for the feature.